README.md

```diff
@@ -230,36 +227,105 @@ You're going to want to stick all of this into your `package.yaml` file.
+this **isn't `polysemy`-specific** problem - basically **all popular effects
+libraries** ended up being bitten by variation of this problem in one way or
+another, resulting in
+[visible drop in performance](https://github.com/lexi-lambda/ghc-proposals/blob/delimited-continuation-primops/proposals/0000-delimited-continuation-primops.md#putting-numbers-to-the-cost)
+compared to equivalent code without use of effect systems.
+
+*Why nobody noticed this for such a long time?*
+
+One factor may be that while GHC's optimizer is
+very, very good in general in optimizing all sorts of abstraction, it's
+relatively complex and hard to predict - authors of libraries may have not
+deemed location of code relevant, even though it had big effect at the end.
+The other is that maybe **it doesn't matter as much** as I we like to tell
+ourselves. Many of these effects
+libraries are used in production and they're doing just fine, because maximum
+performance usually matters in small, controlled areas of code, that often
+don't use features of effect systems at all.
+
+*What can we do about this?*
+
+Luckily, the same person that uncovered this problems proposed a
+[solution](https://github.com/lexi-lambda/ghc-proposals/blob/delimited-continuation-primops/proposals/0000-delimited-continuation-primops.md) -
+set of primops that will allow interpretation of effects at runtime, with
+minimal overhead. It's not *zero-cost* as we hoped for with `polysemy` at
+first, but it should have negligible effect on performance in real life  and
+compared to current solutions, it should be much more predictable and even
+resolve some problems with behaviour of
+[specific effects](https://github.com/polysemy-research/polysemy/issues/246).
+You can try out experimental library that uses proposed features
+[here](https://github.com/hasura/eff).
+
+When it comes to `polysemy`, once GHC proposal lands, we consider option of
+switching to implementation based on it. This will probably require some
+breaking changes, but should resolve performance issues and maybe even make
+implementation of higher-order effects easier.
+
+If you're interested in more details, see
+Alexis King's
+[talk about the problem](https://www.youtube.com/watch?v=0jI-AlWEwYI),
+Sandy Maguire's
+[followup about how it relates to `polysemy`](https://reasonablypolymorphic.com/blog/mea-culpa/) and
+[GHC proposal](https://github.com/ghc-proposals/ghc-proposals/pull/313) that
+adds features needed for new type of implementation.
+
+### TL;DR
+
+Basically all current effects libraries (including `polysemy` and
+even `mtl`) got performance wrong - **but**, there's ongoing work on extending
+GHC with features that will allow for creation of effects implementation with
+stable and strong performance. It's what `polysemy` may choose at some point,
+but it will probably require few breaking changes.
+
+## Stellar Engineering - Aligning the stars to optimize `polysemy` ~~away~~ better
```

Comment:
Perhaps we could mention `-fexponse-all-unfoldings` as a last resort, which afaiu is one of the "barriers" required for getting zero-cost effects?

Reply:
If I remember right, we were talking about dropping `loopbreaker` (which should be needed to make this work) because it's not going to be useful in general anyway and it's an ugly hack - I actually wonder how much of those flags are still going to make any difference in that case. And even if we kept supporting it, I'm worried that if we tell people to enable `-fexpose-all-unfoldings` for small projects, they will do so just to later find out that their project grew too big to compile with it.

Reply:
let's just remove this entire section

Reply:
Okay - maybe we can start another issue for creating new set of flags that is known to somewhat help in absence of `loopbreaker`.